### ISSUE TITLE
`UnmarshalStrict()` doesn't error on duplicate keys

### ISSUE THREAD

**Issue by aeijdenberg** [go-yaml/yaml#284]
I spent yesterday afternoon tracking down a bug in our Concourse configuration whereby the root cause was a YAML file containing:

```yaml
resource_types:
- name: foo
- name: bar

... hundreds of unrelated lines later ...

resource_types:
- name: baz
```

where the final `resource_types` effectively deleted the first.

I thought it would be within the spirit of what `UnmarshalStrict()` sounds like it should do, if it would return an error if a key is attempted to be set in a map (or struct) that already has value.

PR forthcoming for your consideration...

**Comment by aeijdenberg** [comment]
Related: <https://github.com/go-yaml/yaml/issues/154>

**Comment by xunwu** [comment]
I am facing the same issue here, I think it should throw an error if duplicate keys are found, rather than silently replacing the value

**Comment by rogpeppe** [comment]
Fixed by #307.
